sycl/source/detail/graph_memory_pool.hpp

```diff
@@ -0,0 +1,209 @@
+    }
+
+    default:
+      throw sycl::exception(sycl::make_error_code(errc::feature_not_supported),
+                            "Only device allocations are currently supported "
+                            "in graph allocation nodes!");
+      break;
+    }
+
+    return Alloc;
+  }
+
+  /// Return the total amount of memory being used by this pool
+  size_t getMemUseCurrent() const {
+    size_t TotalMem = 0;
+    for (auto &PhysicalMem : MPhysicalMem) {
+      TotalMem += PhysicalMem->size();
+    }
+
+    return TotalMem;
+  }
+
+  /// For device allocations using virtual memory this function allocates
+  /// physical memory and maps each virtual range to it, should be called during
+  /// graph finalization.
+  void allocateAndMapAll() {
+    // Queue used for any initializing of memory, zero-init etc.
+    sycl::queue Queue{MContext, MDevice};
+
+    for (auto &Allocation : MAllocations) {
+      // Set access mode
+      void *Ptr = Allocation.first;
+      alloc_info &AllocInfo = Allocation.second;
+      address_access_mode AccessMode = AllocInfo.ReadOnly
+                                           ? address_access_mode::read
+                                           : address_access_mode::read_write;
+
+      // Create physical memory
+      auto PhysicalMem = std::make_shared<physical_mem_impl>(MDevice, MContext,
+                                                             AllocInfo.Size);
+      // Map the virtual reservation to it
+      PhysicalMem->map(reinterpret_cast<uintptr_t>(Ptr), AllocInfo.Size,
+                       AccessMode, 0);
+
+      // Zero init if required
+      if (AllocInfo.ZeroInit) {
+        Queue.submit(
+            [&](sycl::handler &CGH) { CGH.memset(Ptr, 0, AllocInfo.Size); });
+      }
+
+      MPhysicalMem.push_back(PhysicalMem);
+      AllocInfo.PhysicalMemID = MPhysicalMem.size() - 1;
+      AllocInfo.Mapped = true;
+    }
+
+    // Wait on any operations we enqueued.
+    Queue.wait_and_throw();
+  }
+
+  /// For device virtual reservations unmap and deallocate physical memory.
+  /// Virtual reservations are not released and can be reallocated/mapped again.
+  /// Typically called on executable graph destruction.
+  void deallocateAndUnmapAll() {
+    for (auto &[Ptr, AllocInfo] : MAllocations) {
+      // Unmap allocations before physical memory is released
+      // Physical mem is released when MPhysicalMem is cleared
+      unmap(Ptr, AllocInfo.Size, MContext);
+      AllocInfo.PhysicalMemID = 0;
+      AllocInfo.Mapped = false;
+    }
+
+    MPhysicalMem.clear();
+  }
+
+  /// True if this pool has any allocations
+  bool hasAllocations() const { return MAllocations.size(); }
+
+  // True if an allocation exists for this pointer
+  bool hasAllocation(void *Ptr) const {
+    return MAllocations.find(Ptr) != MAllocations.end();
+  }
+
+private:
+  /// Returns an aligned byte size given a required granularity
+  size_t alignByteSize(size_t UnalignedByteSize, size_t Granularity) {
```

Comment:
Can this be static? To be pedantic, this could also use doxygen comments for the params.

Reply:
Done!